src/v/storage/tests/storage_e2e_test.cc

```diff
@@ -35,6 +35,10 @@
 #include <iterator>
 #include <numeric>
 
+storage::disk_log_impl* get_disk_log(storage::log log) {
+    return reinterpret_cast<storage::disk_log_impl*>(log.get_impl());
```

Comment:
The upcast can be done using `dynamic_cast` operator. The `reinterpret_cast` here breaks aliasing rules and might be causing an undefined behavior.